src/query_plan/operation.rs

```diff
@@ -1124,6 +1281,40 @@ pub(crate) mod normalized_inline_fragment_selection {
                 Ok(None)
             }
         }
+
+        pub(crate) fn can_add_to(
+            &self,
+            parent_type: &CompositeTypeDefinitionPosition,
+            schema: &ValidFederationSchema,
+        ) -> Result<bool, FederationError> {
```

Comment:
Remaining question: I'm not sure how to port this from JS. Inheritance made this hard to parse out. A pointer on where to look would be helpful.

Reply:
As per my other comment -> `can_add_to` methods should be on the corresponding top level (outside of inner mod) `XyzSelection` impls. 

* `NormalizedFragmentSpreadSelection` ->[ should always return `true`](https://github.com/apollographql/federation/blob/main/internals-js/src/operations.ts#L3689-L3693)
* `InlineFragmentSelection` -> [checks if casted type can be added and then validates the selection set](https://github.com/apollographql/federation/blob/main/internals-js/src/operations.ts#L3330-L3344)
* `FieldSelection` -> you already got it implemented ([JS logic](https://github.com/apollographql/federation/blob/main/internals-js/src/operations.ts#L3091-L3107)